### PR TITLE
[Editor] [sc-2377] The Remove Format feature is not removing all types of extra formatting

### PR DESCRIPTION
**Pull request by NicholasJarr**
## Describe the changes in this pull request

This PR changes the schema to allow `htmlAttributes` to be removed using the **Remove Format** feature.

App PR: https://github.com/hajrice/helpjuice4/pull/3099

#### Before

https://github.com/Helpjuice/ckeditor5/assets/28270050/c0541047-3bd2-46b1-b0f6-2c023b97619d

#### After

https://github.com/Helpjuice/ckeditor5/assets/28270050/24724eed-e83e-4bba-afa4-e6bd21a78c4e

### Does this resolve an existing issue?

[[sc-2377](https://app.shortcut.com/helpjuice/story/2377/the-remove-format-feature-is-not-removing-all-types-of-extra-formatting)]

### How should this be tested during the review?

Insert the following HTML through the source editor on the editor preview (I got this HTML from copying the content from this [file](https://github.com/Helpjuice/ckeditor5/files/12030630/Test.docx) and pasting with formatting in the editor; I just removed some unnecessary spacing and tags):

```
<h1>
    Test
</h1>
<h2>
    Section 1
</h2>
<p class="MsoNormal" style="text-indent:.5in;">
    A paragraph with some random thoughts.
</p>
<h2>
    Section 2
</h2>
<p class="MsoNormal">
    Another paragraph
</p>
```

Now, select everything and click on **Remove Format** (the eraser icon in the toolbar). It will remove all the custom classes and styles (you can check the source editor to verify this).

### CODE REVIEW

**Comment by Bondok6** [comment]
After testing, it found that it works fine on all elements except links (`<a>` tags)

TODO: Find a way to remove all the Style attribute from `<a>` tags.


> I've been trying to find a way to remove style/formatting from Links but CKeditor doesn't seem to let us do that.

> [Docs:](https://ckeditor.com/docs/ckeditor5/latest/features/remove-format.html) This feature does not remove block-level formatting semantic data ([links](https://ckeditor.com/docs/ckeditor5/latest/features/link.html)).

Screenshot from 2023-07-19 08-59-08

But I found that I could remove the entire link and its style/format and represent it as text without formatting
( Convert the link to plain text )

Example)

> Before clicking on the format button

```html
<p style="color: blue;">plain text <a style="color:red; background:green" href="www.google.com">Google</a> <p>
```

> After clicking on the format button

```html
<p> plain text Google <p>
```

The user will have to add the link again manually in this case.

**Comment by Bondok6** [comment]
Now, we have two options, I'll check with Amanda about which option would work best.

Option 1) Remove all formatting from all elements, except links `<a>` tags.
Option 2) Remove all formatting from all elements, and convert the link to plain text.

Hey @NicholasJarr @EmirVatric , Do you have any suggestions for alternative approaches that I could try? Your input would be greatly appreciated. Thank you!

> Additionally, I can create a separate card for addressing the link formatting problem.

**Comment by EmirVatric** [comment]
> 

Is there an option to keep the `a` tag with `href` that they choose rather than converting it to a plain text?

**Comment by Bondok6** [comment]
> > 
> 
> Is there an option to keep the `a` tag with `href` that they choose rather than converting it to a plain text?

@EmirVatric, As far as I know the CKeditor doesn't have a built-in feature or option for that, but I can try to make a custom plugin or something similar to achieve that ( Not 100% sure if this will work )

**Comment by NicholasJarr** [comment]
> Now, we have two options, I'll check with Amanda about which option would work best.
> 
> Option 1) Remove all formatting from all elements, except links `<a>` tags.
> Option 2) Remove all formatting from all elements, and convert the link to plain text.

I agree with Emir and I think we should keep the `<a>`. We should only remove the styles and classes. I tried it on the editor (in production) and it worked. I added a link, made it bold, added a background color, and then I removed the formatting from it. The link stayed the same and the styles were gone. Looking at the source editor, it worked because the styling was in the text inside the `<a>` tag not the tag itself.

If I understood correctly, the problem is when the attributes are set in the link tag. Is that it?

**Comment by Bondok6** [comment]
> > Now, we have two options, I'll check with Amanda about which option would work best.
> > Option 1) Remove all formatting from all elements, except links `<a>` tags.
> > Option 2) Remove all formatting from all elements, and convert the link to plain text.
> 
> I agree with Emir and I think we should keep the `<a>`. We should only remove the styles and classes. I tried it on the editor (in production) and it worked. I added a link, made it bold, added a background color, and then I removed the formatting from it. The link stayed the same and the styles were gone. Looking at the source editor, it worked because the styling was in the text inside the `<a>` tag not the tag itself.
> 
> If I understood correctly, the problem is when the attributes are set in the link tag. Is that it?

@NicholasJarr,

Yes it works if you added a link then changed the format through the CKeditor,
But The problem occurs if you copy a link from anywhere and paste it in CKeditor and that link has a style we won't be able to remove that style with the remove formatting feature

Also, Try to add `style="color:red;"` manually to the link through the Source we won't be able to remove that style with the remove formatting feature

[Screencast from 19 يول, 2023 EEST 03:38:19 م.webm](https://github.com/Helpjuice/ckeditor5/assets/45231276/3b9ea48b-e098-4251-95b0-855d393f7884)

**Comment by NicholasJarr** [comment]
> Yes it works if you added a link then changed the format through the CKeditor,
> But The problem occurs if you copy a link from anywhere and paste it in CKeditor and that link has a style we won't be able to remove that style with the remove formatting feature
> 
> Also, Try to add `style="color:red;"` manually to the link through the Source we won't be able to remove that style with the remove formatting feature

I understand now. If possible, we should find a way for it to work on links too. Does Remove Format command have a check to skip links?

**Comment by Bondok6** [comment]
> > Yes it works if you added a link then changed the format through the CKeditor,
> > But The problem occurs if you copy a link from anywhere and paste it in CKeditor and that link has a style we won't be able to remove that style with the remove formatting feature
> > Also, Try to add `style="color:red;"` manually to the link through the Source we won't be able to remove that style with the remove formatting feature
> 
> I understand now. If possible, we should find a way for it to work on links too. Does Remove Format command have a check to skip links?

The links seem to be skipped somehow, but I checked the code here

https://github.com/Helpjuice/ckeditor5/blob/helpjuice-build-v38.0.0/packages/ckeditor5-remove-format/src/removeformatcommand.ts

But I couldn't know how they skipped it

**Comment by Bondok6** [comment]
I also noticed that when adding color to a link using CKEditor, the editor creates a span element, applies the style to it, and then appends it to the a tag, So when removing the style using removing format feature it works because it remove the style from the span ( That's why it works when applying the style using Ckeditor )

**Comment by NicholasJarr** [comment]
> The links seem to be skipped somehow, but I checked the code here
> 
> https://github.com/Helpjuice/ckeditor5/blob/helpjuice-build-v38.0.0/packages/ckeditor5-remove-format/src/removeformatcommand.ts
> 
> But I couldn't know how they skipped it

I checked the code and I think I found the reason: `a` tags use a different attribute to store `class` and `style` values. Regularly, elements use `htmlAttributes`, but `a` elements use the `htmlA` attribute instead. Check these links for more information:

https://github.com/Helpjuice/ckeditor5/blob/7ff4a243254e8a1b6e0a2190de9d8153b1b40242/packages/ckeditor5-html-support/src/schemadefinitions.ts#L616-L624

https://github.com/Helpjuice/ckeditor5/blob/5d987420d65903f2118d889ed140e54943074f0e/packages/ckeditor5-style/tests/integrations/link.js#L351-L357

Adding  another call to `extend` using `htmlA` should fix this issue. However, there's one caveat (this includes the implementation for regular elements), it will remove `data-` attributes from the elements too. Not sure if that's a big deal, but it's something we should consider.

**Comment by Bondok6** [comment]
> > The links seem to be skipped somehow, but I checked the code here
> > https://github.com/Helpjuice/ckeditor5/blob/helpjuice-build-v38.0.0/packages/ckeditor5-remove-format/src/removeformatcommand.ts
> > But I couldn't know how they skipped it
> 
> I checked the code and I think I found the reason: `a` tags use a different attribute to store `class` and `style` values. Regularly, elements use `htmlAttributes`, but `a` elements use the `htmlA` attribute instead. Check these links for more information:
> 
> https://github.com/Helpjuice/ckeditor5/blob/7ff4a243254e8a1b6e0a2190de9d8153b1b40242/packages/ckeditor5-html-support/src/schemadefinitions.ts#L616-L624
> 
> https://github.com/Helpjuice/ckeditor5/blob/5d987420d65903f2118d889ed140e54943074f0e/packages/ckeditor5-style/tests/integrations/link.js#L351-L357
> 
> Adding another call to `extend` using `htmlA` should fix this issue. However, there's one caveat (this includes the implementation for regular elements), it will remove `data-` attributes from the elements too. Not sure if that's a big deal, but it's something we should consider.


Thanks @NicholasJarr, I understand now why it was not working with Links, I just added another call to extend using `htmlA` and It works as expected.

```ts

editor.model.schema.setAttributeProperties( 'htmlA', { isFormatting: true } );

```

Regarding the 'data-' attributes, I think it's ok if they are removed.

As far as I know, we're creating the 'data-' attributes that CK needs like `data-toc`, so It won't be removed

> I test it here

[Screencast from 20 يول, 2023 EEST 07:17:28 ص.webm](https://github.com/Helpjuice/ckeditor5/assets/45231276/6ecfc4cf-ffec-48ec-bd7d-67e89c76197e)

**Comment by NicholasJarr** [comment]
> Regarding the 'data-' attributes, I think it's ok if they are removed.
> 
> As far as I know, we're creating the 'data-' attributes that CK needs like `data-toc`, so It won't be removed
> 
> I test it here

Ok. What about the other options like Open Link in New Tab, Downloadable? Are they removed?

**Comment by Bondok6** [comment]
> > Regarding the 'data-' attributes, I think it's ok if they are removed.
> > As far as I know, we're creating the 'data-' attributes that CK needs like `data-toc`, so It won't be removed
> > I test it here
> 
> Ok. What about the other options like Open Link in New Tab, Downloadable? Are they removed?

Hey @NicholasJarr , I test them, they are not removed

[Screencast from 21 يول, 2023 EEST 04:46:59 م.webm](https://github.com/Helpjuice/ckeditor5/assets/45231276/c34f2b6c-45c5-4a16-b81b-e3dde0ac616c)

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts [hunk 1]
```diff
@@ -150,7 +150,7 @@ export default class LinkCommand extends Command {
 	 * @param href Link destination.
 	 * @param manualDecoratorIds The information about manual decorator attributes to be applied or removed upon execution.
 	 */
-	public override execute( href: string, text: string, manualDecoratorIds: Record<string, boolean> = {} ): void {
+	public override execute( href: string, text: string | null = null, manualDecoratorIds: Record<string, boolean> = {} ): void {
```
Comment (NicholasJarr) [inline packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts:153]: Why do we need this change?

**Review by NicholasJarr — COMMENTED** [review]
Looks good to me 👍🏻 . Before approving, I just have a comment and a question: why all of these type definitions files were removed?

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts (hunk 1 above)
Comment (Bondok6) [reply packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts:153]: After rebasing, I get two TS errors when running `npm run build`
( from `helpjuice-build-v38.0.0` branch)

> Adding an optional value to the text parameter fixed the issue for me

Screenshot from 2023-07-20 16-30-40

**Comment by Bondok6** [comment]
> Looks good to me 👍🏻 . Before approving, I just have a comment and a question: why all of these type definitions files were removed?

@NicholasJarr,

The `build` directory in the `helpjuice-build-v38.0.0` branch is not updated, that's why we've here all of these types of definitions removed.

> I Checked that by switching to `helpjuice-build-v38.0.0` branch, pulling the latest changes, removing the build directory `rm -rf build`, and running `npm run build`

Maybe @SpaYco didn't remove the old build folder before running `npm run build` and that's why the type definitions weren't removed in  `helpjuice-build-v38.0.0` branch.

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts (hunk 1 above)
Comment (NicholasJarr) [reply packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts:153]: @SpaYco

Any thoughts on this? Looks like it's related to the links PR

**Comment by NicholasJarr** [comment]
> The `build` directory in the `helpjuice-build-v38.0.0` branch is not updated, that's why we've here all of these types of definitions removed.

Ok. Makes sense.

> Maybe @SpaYco didn't remove the old build folder before running `npm run build` and that's why the type definitions weren't removed in `helpjuice-build-v38.0.0` branch.

@SpaYco, can you confirm this?

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts (hunk 1 above)
Comment (SpaYco) [reply packages/ckeditor5-build-helpjuice-v2/src/plugins/link/src/linkcommand.ts:153]: @NicholasJarr looks like CKFinder and CKBox are throwing those errors after my changes to links plugin, the problem is that i didn't get those errors on my end when i build the app, but @Bondok6 receives them. i'm not sure we use those plugins, but it's okay to have the text parameter optional as I have fail safe to use `href` if `text` is null.

**Comment by SpaYco** [comment]
@NicholasJarr @Bondok6 No i haven't removed the build folder, i wasn't aware that i should do that, is there a reason why we need to do that?

**Comment by Bondok6** [comment]
> @NicholasJarr @Bondok6 No i haven't removed the build folder, i wasn't aware that i should do that, is there a reason why we need to do that?

@SpaYco 

I checked the previous commits and I found in this commit https://github.com/Helpjuice/ckeditor5/commit/5e9562f0d971836f53f620edc9d449a42809e5a2

You removed the `tests` and `docs` folders ( useless files ), and their Type definitions were removed as well in the build folder.

And the next commit here https://github.com/Helpjuice/ckeditor5/commit/9c49e2e92236bd46e65b9e4e20cee1db6db916b6

The type definitions were added again in the build folder for some reason.

So now in the `helpjuice-build-v38.0.0` branch, we have those Type definitions in the build folder and I think we don't need them.

That's why this PR removes those types of definitions files when I run `npm run build`

I think it's ok to remove these files because we already removed tests and docs folders ( useless files )

**Comment by SpaYco** [comment]
@Bondok6 we need to figure out why the builds show different results for us.

i also agree, ee don't need the type definitions, we're only using hjeditor.js.

**Comment by NicholasJarr** [comment]
> @Bondok6 we need to figure out why the builds show different results for us.

@SpaYco, maybe it's the Node version? Can you try following the steps [here](https://github.com/hajrice/helpjuice4/wiki/How-to-develop-for-CKEditor) (the `nvm` part) and try rebuilding the editor locally?

**Review by NicholasJarr — COMMENTED** [review]
@Bondok6

Looks good to me 👍🏻 . I can't approve it in GH because it doesn't let me approve a PR that I'm an author. @SpaYco, can you take a look too?

**Review by Bondok6 — APPROVED** [review]

**Comment by Bondok6** [comment]
Hey guys,

if you don't mind I'll merge this PR because I need to open some other PRs and that's led to having some conflicts with them and with `hjeditor.js` file in helpjuice repo.

Feel free to leave any comments, if there's anything in this PR.

**Pull request merged**